File: gpu/Electron/src/main/telemetry.ts
import fs from 'fs';
import path from 'path';
import { config } from './config';

export interface FpsStreamSeconds {
  acceptable: {
    '25_to_30_fps': number;
    '20_to_24_fps': number;
  };
  unacceptable: {
    '10_to_19_fps': number;
    '5_to_9_fps': number;
    'under_5_fps': number;
  };
}

export interface FpsMetricsPayload {
  timestamp: string;
  machine_id: string;
  framework: string;
  hardware_mode: string;
  window_duration_seconds: number;
  active_streams: number;
  fps_stream_seconds: FpsStreamSeconds;
}

export class TelemetryManager {
  private currentBuckets: FpsStreamSeconds;
  private tickCountInWindow: number = 0;
  private totalFlushes: number = 0;
  private logFilePath: string;
  private activeStreams: number;
  private listeners: ((payload: FpsMetricsPayload, currentWindowSec: number, streamFpsList: number[]) => void)[] = [];

  constructor() {
    this.activeStreams = config.streamCount;
    this.currentBuckets = this.createEmptyBuckets();
    this.logFilePath = this.resolveLogFilePath(config.fpsLogPath);
  }

  private createEmptyBuckets(): FpsStreamSeconds {
    return {
      acceptable: {
        '25_to_30_fps': 0,
        '20_to_24_fps': 0,
      },
      unacceptable: {
        '10_to_19_fps': 0,
        '5_to_9_fps': 0,
        'under_5_fps': 0,
      },
    };
  }

  private resolveLogFilePath(targetPath: string): string {
    const dir = path.dirname(targetPath);
    try {
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      fs.accessSync(dir, fs.constants.W_OK);
      return targetPath;
    } catch (err) {
      console.warn(`[Telemetry] Cannot write to ${targetPath} (${(err as Error).message}). Falling back to local ./logs directory.`);
      const fallbackDir = path.resolve(process.cwd(), 'logs');
      if (!fs.existsSync(fallbackDir)) {
        fs.mkdirSync(fallbackDir, { recursive: true });
      }
      return path.join(fallbackDir, 'fps_metrics.log');
    }
  }

  /**
   * Called on every 1-second tick with the frame counts painted by each stream.
   * @param streamFpsList Array of FPS painted during the past second for each stream.
   */
  public recordTick(streamFpsList: number[]): void {
    this.tickCountInWindow++;
    this.activeStreams = streamFpsList.length;

    for (const fps of streamFpsList) {
      if (fps >= 25) {
        this.currentBuckets.acceptable['25_to_30_fps']++;
      } else if (fps >= 20) {
        this.currentBuckets.acceptable['20_to_24_fps']++;
      } else if (fps >= 10) {
        this.currentBuckets.unacceptable['10_to_19_fps']++;
      } else if (fps >= 5) {
        this.currentBuckets.unacceptable['5_to_9_fps']++;
      } else {
        this.currentBuckets.unacceptable['under_5_fps']++;
      }
    }

    const currentPayload = this.buildPayload();
    this.notifyListeners(currentPayload, this.tickCountInWindow, streamFpsList);

    if (this.tickCountInWindow >= config.windowDurationSeconds) {
      this.flushToDisk();
    }
  }

  public flushToDisk(): void {
    const payload = this.buildPayload();
    const jsonString = JSON.stringify(payload, null, 2);

    try {
      // Append to the log file so continuous 24h benchmark records all 60s windows
      fs.appendFileSync(this.logFilePath, jsonString + '\n\n', 'utf-8');
      this.totalFlushes++;
      console.log(`[Telemetry] Flushed 60s window #${this.totalFlushes} to ${this.logFilePath}`);
      console.log(`[Telemetry] Stats (Mode: ${config.hardwareMode}): Acceptable (25-30: ${payload.fps_stream_seconds.acceptable['25_to_30_fps']}, 20-24: ${payload.fps_stream_seconds.acceptable['20_to_24_fps']}), Unacceptable (<5: ${payload.fps_stream_seconds.unacceptable['under_5_fps']})`);
    } catch (err) {
      console.error(`[Telemetry] Error writing metrics to ${this.logFilePath}:`, err);
    }

    // Immediately reset internal counters to zero for the next window
    this.currentBuckets = this.createEmptyBuckets();
    this.tickCountInWindow = 0;
  }

  public buildPayload(): FpsMetricsPayload {
    return {
      timestamp: new Date().toISOString(),
      machine_id: config.machineId,
      framework: config.framework,
      hardware_mode: config.hardwareMode,
      window_duration_seconds: config.windowDurationSeconds,
      active_streams: this.activeStreams,
      fps_stream_seconds: {
        acceptable: { ...this.currentBuckets.acceptable },
        unacceptable: { ...this.currentBuckets.unacceptable },
      },
    };
  }

  public onUpdate(fn: (payload: FpsMetricsPayload, currentWindowSec: number, streamFpsList: number[]) => void): void {
    this.listeners.push(fn);
  }

  private notifyListeners(payload: FpsMetricsPayload, currentWindowSec: number, streamFpsList: number[]): void {
    for (const listener of this.listeners) {
      try {
        listener(payload, currentWindowSec, streamFpsList);
      } catch (err) {
        console.error('[Telemetry] Listener error:', err);
      }
    }
  }

  public getLogPath(): string {
    return this.logFilePath;
  }
}

export const telemetry = new TelemetryManager();

File: gpu/Rust-Iced/src/telemetry.rs
use chrono::Utc;
use serde::{Deserialize, Serialize};
use std::fs::OpenOptions;
use std::io::Write;
use std::path::PathBuf;
use crate::config::BenchmarkConfig;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AcceptableBuckets {
    #[serde(rename = "25_to_30_fps")]
    pub fps_25_to_30: u32,
    #[serde(rename = "20_to_24_fps")]
    pub fps_20_to_24: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct UnacceptableBuckets {
    #[serde(rename = "10_to_19_fps")]
    pub fps_10_to_19: u32,
    #[serde(rename = "5_to_9_fps")]
    pub fps_5_to_9: u32,
    #[serde(rename = "under_5_fps")]
    pub under_5_fps: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct FpsStreamSeconds {
    pub acceptable: AcceptableBuckets,
    pub unacceptable: UnacceptableBuckets,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FpsMetricsPayload {
    pub timestamp: String,
    pub machine_id: String,
    pub framework: String,
    pub hardware_mode: String,
    pub window_duration_seconds: u32,
    pub active_streams: usize,
    pub fps_stream_seconds: FpsStreamSeconds,
}

pub struct TelemetryManager {
    config: BenchmarkConfig,
    log_path: PathBuf,
    current_buckets: FpsStreamSeconds,
    tick_count_in_window: u32,
    total_flushes: u64,
    active_streams: usize,
    latest_stream_fps: Vec<u32>,
}

impl TelemetryManager {
    pub fn new(config: BenchmarkConfig) -> Self {
        let log_path = config.resolve_log_path();
        let active_streams = config.stream_count;

        Self {
            config,
            log_path,
            current_buckets: FpsStreamSeconds::default(),
            tick_count_in_window: 0,
            total_flushes: 0,
            active_streams,
            latest_stream_fps: vec![0; active_streams],
        }
    }

    pub fn get_log_path(&self) -> String {
        self.log_path.to_string_lossy().to_string()
    }

    #[allow(dead_code)]
    pub fn get_tick_in_window(&self) -> u32 {
        self.tick_count_in_window
    }

    pub fn get_total_flushes(&self) -> u64 {
        self.total_flushes
    }

    #[allow(dead_code)]
    pub fn get_current_buckets(&self) -> &FpsStreamSeconds {
        &self.current_buckets
    }

    #[allow(dead_code)]
    pub fn get_latest_stream_fps(&self) -> &[u32] {
        &self.latest_stream_fps
    }

    pub fn record_tick(&mut self, stream_fps_list: &[u32]) -> (FpsMetricsPayload, u32) {
        self.tick_count_in_window += 1;
        self.active_streams = stream_fps_list.len();
        self.latest_stream_fps = stream_fps_list.to_vec();

        for &fps in stream_fps_list {
            if fps >= 25 {
                self.current_buckets.acceptable.fps_25_to_30 += 1;
            } else if fps >= 20 {
                self.current_buckets.acceptable.fps_20_to_24 += 1;
            } else if fps >= 10 {
                self.current_buckets.unacceptable.fps_10_to_19 += 1;
            } else if fps >= 5 {
                self.current_buckets.unacceptable.fps_5_to_9 += 1;
            } else {
                self.current_buckets.unacceptable.under_5_fps += 1;
            }
        }

        let payload = self.build_payload();
        let current_sec = self.tick_count_in_window;

        if self.tick_count_in_window >= self.config.window_duration_seconds {
            self.flush_to_disk(&payload);
            self.current_buckets = FpsStreamSeconds::default();
            self.tick_count_in_window = 0;
        }

        (payload, current_sec)
    }

    pub fn build_payload(&self) -> FpsMetricsPayload {
        FpsMetricsPayload {
            timestamp: Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
            machine_id: self.config.machine_id.clone(),
            framework: self.config.framework.clone(),
            hardware_mode: self.config.hardware_mode.clone(),
            window_duration_seconds: self.config.window_duration_seconds,
            active_streams: self.active_streams,
            fps_stream_seconds: self.current_buckets.clone(),
        }
    }

    pub fn flush_to_disk(&mut self, payload: &FpsMetricsPayload) {
        let json_str = match serde_json::to_string_pretty(payload) {
            Ok(s) => s,
            Err(e) => {
                eprintln!("[Telemetry] Failed to serialize metrics: {}", e);
                return;
            }
        };

        match OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.log_path)
        {
            Ok(mut file) => {
                if let Err(e) = writeln!(file, "{}\n", json_str) {
                    eprintln!("[Telemetry] Failed to write to {}: {}", self.log_path.display(), e);
                } else {
                    self.total_flushes += 1;
                    println!(
                        "[Telemetry] Flushed 60s window #{} to {}",
                        self.total_flushes,
                        self.log_path.display()
                    );
                    println!(
                        "[Telemetry] Acceptable (25-30: {}, 20-24: {}), Unacceptable (10-19: {}, 5-9: {}, <5: {})",
                        payload.fps_stream_seconds.acceptable.fps_25_to_30,
                        payload.fps_stream_seconds.acceptable.fps_20_to_24,
                        payload.fps_stream_seconds.unacceptable.fps_10_to_19,
                        payload.fps_stream_seconds.unacceptable.fps_5_to_9,
                        payload.fps_stream_seconds.unacceptable.under_5_fps
                    );
                }
            }
            Err(e) => {
                eprintln!("[Telemetry] Failed to open {}: {}", self.log_path.display(), e);
            }
        }
    }
}
